apps/shared/avifpng.c

```diff
@@ -104,61 +108,132 @@ static avifBool avifCopyRawProfile(const char * profile, size_t profileLength, a
-                case PNG_ITXT_COMPRESSION_NONE:
-                case PNG_ITXT_COMPRESSION_zTXt:
-                    textLength = text->itxt_length;
-                    break;
-#endif
-                case PNG_TEXT_COMPRESSION_NONE:
-                case PNG_TEXT_COMPRESSION_zTXt:
-                default:
-                    textLength = text->text_length;
-                    break;
+    for (png_uint_32 i = 0; (!*ignoreExif || !*ignoreXMP) && (i < numTextChunks); ++i, ++text) {
+        avifRWData metadata = { NULL, 0 };
+        avifBool tagExif00WasRemoved;
+        if (!*ignoreExif && !strcmp(text->key, "Raw profile type exif")) {
+            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+                return AVIF_FALSE;
+            }
+            uint32_t exifTiffHeaderOffset;
+            if (avifExtractExifTiffHeaderOffset(metadata.data, metadata.size, &exifTiffHeaderOffset) != AVIF_RESULT_OK) {
+                avifRWDataFree(&metadata);
+                fprintf(stderr, "Exif extraction failed: missing TIFF header\n");
+                return AVIF_FALSE;
+            }
+            avifRWDataFree(&avif->exif);
+            avif->exif = metadata;
+            *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
+        } else if (!*ignoreXMP && !strcmp(text->key, "Raw profile type xmp")) {
+            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+                return AVIF_FALSE;
+            }
+            if (tagExif00WasRemoved) {
+                avifRWDataFree(&metadata);
+                fprintf(stderr, "XMP extraction failed: XMP payload begins with an Exif tag\n");
+                return AVIF_FALSE;
             }
-            return avifCopyRawProfile(text->text, textLength, &avif->exif);
+            avifRWDataFree(&avif->xmp);
+            avif->xmp = metadata;
+            *ignoreXMP = AVIF_TRUE; // Ignore any other XMP chunk.
+        } else if (!strcmp(text->key, "Raw profile type APP1")) {
+            if (!avifCopyRawProfile(text->text, avifGetPngTextLength(text), &metadata, &tagExif00WasRemoved)) {
+                return AVIF_FALSE;
+            }
+            // Differentiate Exif and XMP by checking whether the chunk begins with a valid TIFF header ("IIx0x42" or "MMx42x0").
+            const uint32_t tiffHeaderSize = 4;
+            uint32_t exifTiffHeaderOffset;
+            if (metadata.size > tiffHeaderSize &&
+                (avifExtractExifTiffHeaderOffset(metadata.data, tiffHeaderSize, &exifTiffHeaderOffset) == AVIF_RESULT_OK)) {
+                if (*ignoreExif) {
+                    avifRWDataFree(&metadata);
+                } else {
+                    avifRWDataFree(&avif->exif);
+                    avif->exif = metadata;
+                    *ignoreExif = AVIF_TRUE; // Ignore any other Exif chunk.
+                }
+            } else if (tagExif00WasRemoved) {
```

Comment:
Again, not quite sure why the extra check if a valid Exif is already found above...

Reply:
We check `tagExif00WasRemoved` when it is *not* a valid Exif. I wanted to:
- avoid outputting an XMP chunk with the heading bytes `Exif\0\0` silently removed,
- call `avifCopyRawProfile()` only once,
- avoid reallocating the `avifRWData metadata` just to add/remove the heading bytes `Exif\0\0`,
- keep the code somewhat short and easy to read.

Feel free to suggest alternatives or to discuss whether these constraints are less important than simplifying the code.

Reply:
I don't feel strongly about this, it was just a bit difficult to follow - as discussed above there really should be little difference in Exif or XMP treatment (you either have some APP1 header or you don't in either case) so this special flag just in the Exif case was throwing me off...